File: QuickStart/GridView/CreateRowInfo/AdvanceFilterControl.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Windows.Forms
Imports Telerik.WinControls.Data
Imports Telerik.WinControls.UI
Imports Telerik.WinControls

Namespace Telerik.Examples.WinControls.GridView.CreateRowInfo
	Partial Public Class AdvanceFilterControl
		Inherits UserControl
#Region "Fields"

		Private gridView_Renamed As RadGridView
		Private advancedFilterDescriptor As New FilterDescriptor(String.Empty, FilterOperator.None, Nothing)

#End Region

#Region "Constructors"

		Public Sub New()
			InitializeComponent()
			AddHandler radDdlColumns.SelectedIndexChanged, AddressOf radDdlColumns_SelectedIndexChanged

		End Sub

		Public Sub New(ByVal template As RadGridView)
			Me.New()
			Me.gridView_Renamed = template
		End Sub

#End Region

#Region "Properties"

		Public Property GridView() As RadGridView
			Get
				Return Me.gridView_Renamed
			End Get
			Set(ByVal value As RadGridView)
				If Me.gridView_Renamed IsNot value Then
					If Me.gridView_Renamed IsNot Nothing Then
						RemoveHandler gridView_Renamed.CustomFiltering, AddressOf GridView_CustomFiltering
					End If

					Me.gridView_Renamed = value

					If Me.gridView_Renamed IsNot Nothing Then
						AddHandler gridView_Renamed.CustomFiltering, AddressOf GridView_CustomFiltering
						Me.UpdateTheme(Me.Controls)
					End If

					Me.InitializeAdvanceView()
				End If
			End Set
		End Property

		Private Sub UpdateTheme(ByVal controls As ControlCollection)
			For Each currentControl As Control In controls
				Dim current As RadControl = TryCast(currentControl, RadControl)
				If current IsNot Nothing Then
					current.ThemeName = Me.gridView_Renamed.ThemeName
					Dim pageView As Telerik.WinControls.UI.RadPageView = TryCast(current, Telerik.WinControls.UI.RadPageView)
					If pageView IsNot Nothing Then
						For Each page As RadPageViewPage In pageView.Pages
							Me.UpdateTheme(page.Controls)
						Next page
					End If
				End If
			Next currentControl
		End Sub

#End Region

#Region "Methods"

		Private Sub InitializeAdvanceView()
			Me.radDdlColumns.DataSource = Me.gridView_Renamed.Columns
			Me.radDdlColumns.ValueMember = "Name"
			Me.radDdlColumns.DisplayMember = "Name"
		End Sub

#End Region

#Region "Event Handlers"

		Private Sub radDdlColumns_SelectedIndexChanged(ByVal sender As Object, ByVal e As Telerik.WinControls.UI.Data.PositionChangedEventArgs)
			Me.radDropDownButtonOperator.Items.Clear()
			Me.rtbValue1.Text = String.Empty
			Me.rtbValue1.Enabled = False

			If Me.radDdlColumns.SelectedItem IsNot Nothing Then
				Dim dataColumn As GridViewDataColumn = TryCast(Me.radDdlColumns.SelectedItem.DataBoundItem, GridViewDataColumn)

				Me.radDropDownButtonOperator.Enabled = True
				Me.radDropDownButtonOperator.Text = "No Filter"

				Me.advancedFilterDescriptor.Operator = FilterOperator.None
				Me.advancedFilterDescriptor.PropertyName = dataColumn.Name

				Dim listContext As List(Of FilterOperationContext) = FilterOperationContext.GetFilterOperations(dataColumn.DataType)
				For Each context As FilterOperationContext In listContext
					Dim menuItem As New RadMenuItem(context.Name, context.Operator)
					AddHandler menuItem.Click, AddressOf menuItem_Click
					Me.radDropDownButtonOperator.Items.Add(menuItem)
				Next context
			Else
				Me.radDropDownButtonOperator.Enabled = False
				Me.radDropDownButtonOperator.Text = String.Empty
			End If
		End Sub

		Private Sub menuItem_Click(ByVal sender As Object, ByVal e As EventArgs)
			Dim menuItem As RadMenuItem = TryCast(sender, RadMenuItem)
			Me.radDropDownButtonOperator.Text = menuItem.Text

			Dim filterOperator As FilterOperator = CType(menuItem.Tag, FilterOperator)
			Me.advancedFilterDescriptor.Operator = filterOperator

			Me.rtbValue1.Enabled = filterOperator <> filterOperator.None AndAlso filterOperator <> filterOperator.IsNull AndAlso filterOperator <> filterOperator.IsNotNull
		End Sub

		Private Sub rtbValue1_TextChanged(ByVal sender As Object, ByVal e As EventArgs) Handles rtbValue1.TextChanged
			If Me.radDdlColumns.SelectedItem IsNot Nothing Then
				Dim value As Object = Nothing

				If (Not String.IsNullOrEmpty(Me.rtbValue1.Text)) Then
					Dim dataColumn As GridViewDataColumn = TryCast(Me.radDdlColumns.SelectedItem.DataBoundItem, GridViewDataColumn)
					value = RadDataConverter.Instance.Parse(dataColumn, rtbValue1.Text)
				End If

				Me.advancedFilterDescriptor.Value = value
			End If
		End Sub

		Private Sub radBtnAdvanceApply_Click(ByVal sender As Object, ByVal e As EventArgs) Handles radBtnAdvanceApply.Click
			Dim tableElement As GridTableElement = Me.GridView.TableElement
			Dim masterTemplate As MasterGridViewTemplate = Me.GridView.MasterTemplate
			tableElement.BeginUpdate()
			Me.GridView.EnableCustomFiltering = False
			masterTemplate.DataView.Refresh()
			tableElement.EndUpdate(False)

			For Each row As GridViewRowInfo In Me.GridView.Rows
				For Each cell As GridViewCellInfo In row.Cells
					cell.Style.Reset()
				Next
			Next

			Me.GridView.FilterDescriptors.BeginUpdate()
			Me.GridView.FilterDescriptors.Clear()
			Dim descriptor As FilterDescriptor = TryCast(Me.advancedFilterDescriptor.Clone(), FilterDescriptor)
			Me.GridView.FilterDescriptors.Add(descriptor)
			Me.GridView.FilterDescriptors.EndUpdate()
		End Sub

		Private Sub radTextBoxSearch_TextChanged(ByVal sender As Object, ByVal e As EventArgs) Handles radTextBoxSearch.TextChanged
			Dim tableElement As GridTableElement = Me.gridView_Renamed.TableElement
			Dim masterTemplate As MasterGridViewTemplate = Me.gridView_Renamed.MasterTemplate

			tableElement.BeginUpdate()
			Me.gridView_Renamed.EnableCustomFiltering = True
			masterTemplate.DataView.Refresh()
			tableElement.EndUpdate(False)

			Dim viewChangedEventInfo As New GridViewEventInfo(KnownEvents.ViewChanged, GridEventType.UI, GridEventDispatchMode.Send)
			Dim args As New DataViewChangedEventArgs(ViewChangedAction.FilteringChanged)
			Dim viewChangedEvent As New GridViewEvent(masterTemplate, masterTemplate.DataView, New Object() {args}, viewChangedEventInfo)
			Me.gridView_Renamed.MasterTemplate.SynchronizationService.DispatchEvent(viewChangedEvent)

			Me.gridView_Renamed.TableElement.Update(GridUINotifyAction.StateChanged)
		End Sub

		Private Sub GridView_CustomFiltering(ByVal sender As Object, ByVal e As GridViewCustomFilteringEventArgs)
			Dim searchPattern As String = Me.radTextBoxSearch.Text
			Dim isEmptyPattern As Boolean = String.IsNullOrEmpty(searchPattern)
			e.Visible = isEmptyPattern

			For Each cell As GridViewCellInfo In e.Row.Cells
				Dim value As String = String.Empty
				Dim cellValue As Object = cell.Value

				If cellValue IsNot Nothing AndAlso (Not Convert.IsDBNull(cellValue)) Then
					value = Convert.ToString(cellValue)
				End If

				Dim contains As Boolean = False

				If (Not isEmptyPattern) Then
					contains = value.Contains(searchPattern)
					e.Visible = e.Visible Or contains
				End If

				If contains Then
					cell.Style.CustomizeFill = True
					cell.Style.BackColor = Color.FromArgb(201, 252, 254)
				Else
					cell.Style.Reset()
				End If
			Next cell
		End Sub

#End Region
	End Class
End Namespace

File: QuickStart/GridView/CreateRowInfo/AdvanceFilterControl.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using Telerik.WinControls;
using Telerik.WinControls.Data;
using Telerik.WinControls.UI;

namespace Telerik.Examples.WinControls.GridView.CreateRowInfo
{
    public partial class AdvanceFilterControl : UserControl
    {
        #region Fields

        private RadGridView gridView;
        private FilterDescriptor advancedFilterDescriptor = new FilterDescriptor(String.Empty, FilterOperator.None, null);

        #endregion

        #region Constructors

        public AdvanceFilterControl()
        {
            InitializeComponent();
            this.radDdlColumns.SelectedIndexChanged += new Telerik.WinControls.UI.Data.PositionChangedEventHandler(radDdlColumns_SelectedIndexChanged);

        }

        public AdvanceFilterControl(RadGridView template)
            : this()
        {
            this.gridView = template;
        }

        #endregion

        #region Properties

        public RadGridView GridView
        {
            get
            {
                return this.gridView;
            }
            set
            {
                if (this.gridView != value)
                {
                    if (this.gridView != null)
                    {
                        this.gridView.CustomFiltering -= new GridViewCustomFilteringEventHandler(GridView_CustomFiltering);
                    }

                    this.gridView = value;

                    if (this.gridView != null)
                    {
                        this.gridView.CustomFiltering += new GridViewCustomFilteringEventHandler(GridView_CustomFiltering);
                        this.UpdateTheme(this.Controls);
                    }

                    this.InitializeAdvanceView();
                }
            }
        }

        private void UpdateTheme(ControlCollection controls)
        {
            foreach (Control currentControl in controls)
            {
                RadControl current = currentControl as RadControl;
                if (current != null)
                {
                    current.ThemeName = this.gridView.ThemeName;
                    Telerik.WinControls.UI.RadPageView pageView = current as Telerik.WinControls.UI.RadPageView;
                    if (pageView != null)
                    {
                        foreach (RadPageViewPage page in pageView.Pages)
                        {
                            this.UpdateTheme(page.Controls);
                        }
                    }
                }
            }
        }

        #endregion

        #region Methods

        private void InitializeAdvanceView()
        {
            this.radDdlColumns.DataSource = this.gridView.Columns;
            this.radDdlColumns.ValueMember = "Name";
            this.radDdlColumns.DisplayMember = "Name";
        }

        #endregion

        #region Event Handlers

        private void radDdlColumns_SelectedIndexChanged(object sender, Telerik.WinControls.UI.Data.PositionChangedEventArgs e)
        {
            this.radDropDownButtonOperator.Items.Clear();
            this.rtbValue1.Text = String.Empty;
            this.rtbValue1.Enabled = false;

            if (this.radDdlColumns.SelectedItem != null)
            {
                GridViewDataColumn dataColumn = this.radDdlColumns.SelectedItem.DataBoundItem as GridViewDataColumn;

                this.radDropDownButtonOperator.Enabled = true;
                this.radDropDownButtonOperator.Text = "No Filter";

                this.advancedFilterDescriptor.Operator = FilterOperator.None;
                this.advancedFilterDescriptor.PropertyName = dataColumn.Name;

                List<FilterOperationContext> listContext = FilterOperationContext.GetFilterOperations(dataColumn.DataType);
                foreach (FilterOperationContext context in listContext)
                {
                    RadMenuItem menuItem = new RadMenuItem(context.Name, context.Operator);
                    menuItem.Click += new EventHandler(menuItem_Click);
                    this.radDropDownButtonOperator.Items.Add(menuItem);
                }
            }
            else
            {
                this.radDropDownButtonOperator.Enabled = false;
                this.radDropDownButtonOperator.Text = String.Empty;
            }
        }

        private void menuItem_Click(object sender, EventArgs e)
        {
            RadMenuItem menuItem = sender as RadMenuItem;
            this.radDropDownButtonOperator.Text = menuItem.Text;

            FilterOperator filterOperator = (FilterOperator)menuItem.Tag;
            this.advancedFilterDescriptor.Operator = filterOperator;

            this.rtbValue1.Enabled = filterOperator != FilterOperator.None &&
                                         filterOperator != FilterOperator.IsNull &&
                                         filterOperator != FilterOperator.IsNotNull;
        }

        private void rtbValue1_TextChanged(object sender, EventArgs e)
        {
            if (this.radDdlColumns.SelectedItem != null)
            {
                object value = null;

                if (!String.IsNullOrEmpty(this.rtbValue1.Text))
                {
                    GridViewDataColumn dataColumn = this.radDdlColumns.SelectedItem.DataBoundItem as GridViewDataColumn;
                    value = RadDataConverter.Instance.Parse(dataColumn, rtbValue1.Text);
                }

                this.advancedFilterDescriptor.Value = value;
            }
        }

        private void radBtnAdvanceApply_Click(object sender, EventArgs e)
        {
            this.gridView.EndEdit();

            GridTableElement tableElement = this.gridView.TableElement;
            MasterGridViewTemplate masterTemplate = this.gridView.MasterTemplate;
            tableElement.BeginUpdate();
            this.gridView.EnableCustomFiltering = false;
            masterTemplate.DataView.Refresh();
            tableElement.EndUpdate(false);

            foreach (GridViewRowInfo row in this.gridView.Rows)
            {
                foreach (GridViewCellInfo cell in row.Cells)
                {
                    cell.Style.Reset();
                }
            }

            this.gridView.FilterDescriptors.BeginUpdate();
            this.gridView.FilterDescriptors.Clear();
            FilterDescriptor descriptor = this.advancedFilterDescriptor.Clone() as FilterDescriptor;
            this.gridView.FilterDescriptors.Add(descriptor);
            this.gridView.FilterDescriptors.EndUpdate();
        }

        private void radTextBoxSearch_TextChanged(object sender, EventArgs e)
        {
            this.gridView.EndEdit();

            GridTableElement tableElement = this.gridView.TableElement;
            MasterGridViewTemplate masterTemplate = this.gridView.MasterTemplate;

            tableElement.BeginUpdate();
            this.gridView.FilterDescriptors.Clear();
            this.gridView.EnableCustomFiltering = true;
            masterTemplate.DataView.Refresh();
            tableElement.EndUpdate(false);

            this.gridView.TableElement.ViewElement.UpdateRows(true);
        }

        void GridView_CustomFiltering(object sender, GridViewCustomFilteringEventArgs e)
        {
            string searchPattern = this.radTextBoxSearch.Text; 

            bool isEmptyPattern = String.IsNullOrEmpty(searchPattern);
            e.Visible = isEmptyPattern;

            foreach (GridViewCellInfo cell in e.Row.Cells)
            {
                string value = String.Empty;
                object cellValue = cell.Value;

                if (cellValue != null && !Convert.IsDBNull(cellValue))
                {
                    value = Convert.ToString(cellValue);
                }

                bool contains = false;

                if (!isEmptyPattern)
                {
                    contains = value.Contains(searchPattern);
                    e.Visible |= contains;
                }

                gridView.MasterTemplate.SynchronizationService.SuspendEvent(KnownEvents.RowInvalidated);
                if (contains)
                {
                    cell.Style.CustomizeFill = true;
                    cell.Style.BackColor = Color.FromArgb(201, 252, 254);
                }
                else
                {
                    cell.Style.Reset();
                }
                gridView.MasterTemplate.SynchronizationService.ResumeEvent(KnownEvents.RowInvalidated);
            }
        }

        #endregion
    }
}
